// webdriver-image-comparison/dist/methods/createCompareReport.js
import { resolve as pathResolve } from 'node:path';
import { writeFileSync } from 'node:fs';
export function createCompareReport({ boundingBoxes, data, fileName, folders, size, testContext: { commandName, instanceData: { app, browser, deviceName, isIOS, isMobile, platform, }, framework, parent, tag, title }, }) {
    const { misMatchPercentage, rawMisMatchPercentage } = data;
    const jsonFileName = fileName.split('.').slice(0, -1).join('.');
    const jsonFilePath = pathResolve(folders.actualFolderPath, `${jsonFileName}-report.json`);
    const browserContext = {
        browser,
        platform,
    };
    const mobileContext = {
        ...(app !== 'not-known'
            ? { app }
            : { browser: { name: browser.name, version: isIOS ? platform.version : browser.version } }),
        deviceName,
        platform,
    };
    const jsonData = {
        description: parent,
        test: title,
        tag,
        instanceData: isMobile ? mobileContext : browserContext,
        commandName,
        framework,
        boundingBoxes,
        fileData: {
            actualFilePath: pathResolve(folders.actualFolderPath, fileName),
            baselineFilePath: pathResolve(folders.baselineFolderPath, fileName),
            ...(folders.diffFolderPath && { diffFilePath: pathResolve(folders.diffFolderPath, fileName) }),
            fileName,
            size,
        },
        misMatchPercentage: misMatchPercentage.toString(),
        rawMisMatchPercentage,
    };
    writeFileSync(jsonFilePath, JSON.stringify(jsonData), 'utf8');
}
